_posts/2020-10-21-explicit-term-inference-in-scala-3.md

```diff
@@ -0,0 +1,433 @@
+by adding the import clause 'import scala.language.implicitConversions'
+or by setting the compiler option -language:implicitConversions.
+See the Scala docs for value scala.language.implicitConversions for a discussion
+why the feature should be explicitly enabled.
+```
+The warning can be silenced by adding `import scala.language.implicitConversions`, as
+the message suggests.
+
+### Find out more
+
+You can learn more about this feature in the [documentation](https://dotty.epfl.ch/docs/reference/contextual/conversions.html).
+
+### How was this done in Scala 2?
+
+Scala 2 relied on `implicit` defs and `implicit` function values to implement this pattern:
+
+```scala
+import java.util.Optional
+
+object OptionalConversion {
+  implicit def optionalToOption[A](in: Optional[A]): Option[A] =
+    if (in.isPresent) Some(in.get())
+    else None
+}
+```
+
+which could then be imported:
+```scala
+import OptionalConversion._
+```
+
+Note that at a glance, it is not clear that the definition is intended as an automatic conversion to an expected type:
+ - the name of the method can be a hint, but this is merely a convention that
+ other developers might not share
+ - the implicit def may be intended give you extension methods on its parameter type, depending on its result
+ - there is not straightforward way to verify that this function will not be used as implicit
+ argument for a function with an implicit parameter list
+
+The `Conversion` trait makes the definition explicit and more readable.
+
+
+## A common pattern revisited in Scala 3: Typeclasses
+
+This pattern is fundamental to Scala functional programming libraries such as [Cats](https://typelevel.org/cats/). In Scala 2 we used to heavily rely on `implicit` conversions to add methods
+and on `implicit` parameters to propagate instances, which was a bit cryptic to beginners who were maybe already struggling with new concepts in functional programming.
+
+In Scala 3, this pattern becomes simpler thanks to the new syntax.
+Let's consider a simple typeclass such as `Show` which describes the capability of types to have a `String` representation.
+
+We first describe the interface that all instances of this typeclass should implement:
+
+```scala
+// Show.scala
+trait Show[A]:
+  extension (a: A) def show: String
+```
+We can then add a companion object to provide a couple of auxiliary methods and instances
+that make instance creation less tedious:
+```scala
+// Show.scala
+object Show:
+  def from[A](f: A => String): Show[A] = new Show[A]:
+    extension (a: A) def show: String = f(a)
+
+  given[A: Show] as Show[List[A]] = new Show[List[A]]:
+    extension (ls: List[A]) def show: String = ls.map(_.show).mkString(",")
+```
+
+To use it we need to import the interface and define an instance:
+```scala
+// Main.scala
+import Show.{_, given}
+
+case class Mountain(name: String, height: Int)
+
+given Show[Mountain] =
+  Show.from((m: Mountain) => s"The ${m.name} is ${m.height} meters high")
```

Comment:
This results in `The Mont Blanc is...` which isn't quite right, grammatically, even though `The Matterhorn is...` would be. Maybe just remove `The` and pick a different mountain? ;)

Reply:
Hum, why is that? In French I'd say "Le Mont Blanc ...". Isn't that the same in English?

Reply:
Maybe I can just drop `The`

Reply:
No... for reasons I can't explain. "I'm going to climb Mont Blanc" would be how we'd phrase it. Though if we had a white mountain with an English name, I'm sure we'd say "I'm going to climb _the_ White Mountain". But we also say "_the_ Matterhorn".